keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,504 @@
++------+------+                +------+------+               +------+------+
+       |                              |                             |
+       |           Admit()            |                             |
+       +----------------------------->+                             |
+       |                              |      GetTopologyHint()      |
+       |                              +---------------------------->+
+       |                              |                             |
+       |                              |            Hints            |
+       |                              +<----------------------------+
+       |                              |                             |
+       |                              +-+----------+                |
+       |                              | |          |Hint Merging &  |
+       |                              +-+<---------+Decide Admission|
+       |      admit or reject Pod     |                             |
+       +<-----------------------------+                             |
+       |                              |                             |
+       +                              +                             +
+
+```
+_Figure: The sequence of Pod admission in Kubelet._
+
+Detailedly, to get an information that a resource that container requests is available or not, Topology Manager talks to other managers like CPU Manager and Device Manager, called _hint providers_, which can give the needed information of CPU cores and PCIe peripherals like NIC and GPU, respectively, in each NUMA node. Once collecting all the hints in each NUMA node from the managers, then Topology Manager finally determines which NUMA node's resources in a host can be allocatable for the requested container.
+
+However, the problem raising by this KEP is that there is no hint provider which can give a hint for memory and hugepages resources. That is, Topology Manager could not determine NUMA-aware resource allocation in best way due to absence of memory-related information, which causes inter-node commucation overhead among CPU, memory, and PCIe devices and leads overall performance degradation of containers.
+
+To resolve the problem, Memory Manager is proposed in this KEP to provide a way to guarantee NUMA affinity of memory as well as other resources (i.e., CPU and PCIe peripherals) in a procedure of container deployment.
+
+# Motivation
+
+## Related Features
+
+- [Topology Manager][topology-manager] is a feature that collects topology hints from various hint providers (e.g., CPU Manager and Device Manager) to calculate which NUMA node can give a requested amount of resources for a container. With this calculated hints, Topology Manager also refers to node topology policy (i.e. _burstable, restricted, single-numa-policy_) and NUMA affinity of containers, and then it finally determines if container in a pod can be deployed onto the host machine.
```

Comment:
According to the [Control Topology Management Policies on a node](https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/) document there is no burstable topology policy. Available policies are: none, best-effort, restricted and single-numa-policy. burstable is a QoS class AFAIR.

Reply:
`burstable topology policy`, I think it is just stupid mistake :(

Reply:
I will update later